apps/src/aichat/views/ChatWorkspace.tsx

```diff
@@ -50,12 +70,31 @@ const ChatWorkspace: React.FunctionComponent<ChatWorkspaceProps> = ({
     }
   }, [messagesString, isWaitingForChatResponse]);
 
-  const dispatch = useAppDispatch();
+  useEffect(() => {
+    let selectedStudentName = null;
+    if (viewAsUserId) {
+      const selectedStudent = Object.values(students).find(
+        student => student.id === viewAsUserId
+      );
+      if (selectedStudent) {
+        selectedStudentName = getShortName(selectedStudent.name);
+      }
+    }
+    setSelectedStudentName(selectedStudentName);
+  }, [viewAsUserId, students]);
 
-  const onCloseWarningModal = useCallback(
-    () => dispatch(setShowWarningModal(false)),
-    [dispatch]
-  );
+  useEffect(() => {
+    // If a teacher is viewing workspace as a student when level first loads (user_id param included in url)
+    // or from when viewing workspace as a participant, default to the student chat history view.
+    if (
+      viewAsUserId &&
+      (!viewMode || viewMode === WORKSPACE_VIEW_MODE.PARTICIPANT)
+    ) {
+      setViewMode(WORKSPACE_VIEW_MODE.STUDENT_CHAT_HISTORY);
+    } else if (!viewAsUserId) {
+      setViewMode(WORKSPACE_VIEW_MODE.PARTICIPANT);
+    }
+  }, [viewAsUserId, viewMode]);
```

Comment:
This has the same render trap by setting the local state via a `useEffect`, but the fix is a little trickier since `setViewMode` is _also_ set via a callback. In this case, I think it should probably be in the viewMode initializer.

```
const [viewMode, setViewMode] = useState<string>( () => {
    if (
      viewAsUserId &&
      (!viewMode || viewMode === WORKSPACE_VIEW_MODE.PARTICIPANT)
    ) {
      return WORKSPACE_VIEW_MODE.STUDENT_CHAT_HISTORY;
    } else if (!viewAsUserId) {
      return WORKSPACE_VIEW_MODE.PARTICIPANT;
    }
});
```

With the caveat that this would then _not_ update if the `viewAsUserId` value changes. If that is needed and `viewAsUserId` can change, then it might be simpler to just leave the existing logic in place. I bet it can still be simplified, but I'm not sure how off the top of my head.

Regardless, the conditional in here may also be able to be simplified - since it's always initialized to a value, I dropped the `| null` typing, and I _think_ the `!viewMode ||` part of the conditional can come out, but I wasn't completely sure of that.

Anyway, def check my work and make sure it still works before just blindly dropping it in. 😉

Reply:
Hey Jim - thanks for the comment!

We do need to update if `viewAsUserId` changes. If `viewAsUserId` is `null`, then the teacher is viewing lab as a participant: 'http://localhost-studio.code.org:3000/s/allthethings/lessons/47/levels/1?section_id=4&user_id&viewAs=Instructor' and we don't want to show the two tabs.

I added the `!viewMode || ` art of the conditional so that if the URL is loaded for the first time and the `user_id` param is included in the URL, then we do want to default to the chat history view so I think I'll leave thing as-is.

Reply:
Hi again - I refactored and made thing a lot more simple. Thanks again.